python3Packages.jaxlib-bin: cudatoolkit -> redist packages

samuela opened the pull request:
## Description of changes

Migrate jaxlib-bin to use the CUDA redist packages. In the process, I also extracted the CUDA testing code into its own script with some documentation for running the tests.

I have tested these changes with the new `test_cuda.py` and confirmed that they work as intended.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/test_cuda.py:
You can package this a a `writePython3Bin` exposed in the passthru, for example

samuela replied:
Is GPU access possible in `passthru.tests`?

SomeoneSerge replied:
GPU access isn't possible inside the Nix sandbox unless you relax it, but you can package a script which, when run _outside_ the sandbox our inside a relaxed one, would test jaxlib for the GPU support: `nix run .#python3Packages.jaxlibWithCuda.passthru.test_cuda`.

Later we can also reuse this script to expose a `passthru.tests`-like check with the `requiredSystemFeatures = [ "cuda" ]`

samuela replied:
I just tried to  implement this but ran into an issue with ptxas. The details are available in [this gist](https://gist.github.com/samuela/4f2647dcfe41466586203fba2f2f5d38). In short, jaxlib requires ptxas to be in PATH at run time. We are currently accomplishing this in both the source and binary derivations by symbolically linking it into `$out/bin`:
https://github.com/NixOS/nixpkgs/blob/5e6e6e54a667a183792a855e99905701c58c7650/pkgs/development/python-modules/jaxlib/bin.nix#L219

For reasons I do not yet understand, this is not compatible with `python3.withPackages`. This is a separate issue -- and one that I'd like to resolve -- but absent a quick solution to it, `passthru` does not seem feasible for the time being.

wdyt @SomeoneSerge ?

SomeoneSerge replied:
I suppose `withPackages` only creates a wrapper that runs the interpreter with an extended `PYTHONPATH`, not `PATH`? How does jax(lib) run `ptxas`?

If it expects it on the `PATH` we could probably put `getBin cuda_nvcc` in `propagatedBuildInputs` (not too sure this would work but I expect it to)

We could patch jax(lib)'s sources to run `ptxas` by its absolute path under `getBin cuda_nvcc`

SomeoneSerge replied:
> ...thefollowi ng locations were considered: /nix/store/w48pp30vl0jk2wvwdqgb9zg06my5m7sy-python3-3.11.7-env/lib/python3.11/site-packages/jaxlib/cuda/bin/ptxas

So it doesn't look for ptxas in `$out/bin` but in `$out/${python.sitePackages}/jaxlib/cuda/bin`. You can symlink it there

samuela replied:
ooh good idea! done in https://github.com/NixOS/nixpkgs/pull/288829/commits/0e894e01ca709298a90ae0527daf7534ae76294b

SomeoneSerge commented on pkgs/top-level/python-packages.nix:
```diff
@@ -5800,7 +5800,9 @@ self: super: with self; {
 
-  jax = callPackage ../development/python-modules/jax { };
+  jax = callPackage ../development/python-modules/jax {
+    inherit (pkgs.writers) writePython3Bin;
```
I think it's OK to just take `writers` as argument (there prboably are precedents of doing so. Tbd: check)

samuela replied:
hmm i'm not able to find any precedent for that in `python-packages.nix`. i don't really have a preference though

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/bin.nix:
```diff
@@ -211,12 +206,14 @@ buildPythonPackage {
-    mkdir -p $out/bin
-    ln -s ${cudatoolkit}/bin/ptxas $out/bin/ptxas
+    mkdir -p $out/${python.sitePackages}/jaxlib/cuda/bin
+    ln -s ${cudaPackagesGoogle.cuda_nvcc.bin}/bin/ptxas $out/${python.sitePackages}/jaxlib/cuda/bin/ptxas
```
Nit: let's not constrain our future self and use `getBin`

samuela replied:
done in https://github.com/NixOS/nixpkgs/pull/288829/commits/cdb1b0179d40b3d6c0b398946a5af112ef7ac6f5

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/bin.nix:
```diff
@@ -189,18 +189,13 @@ buildPythonPackage {
-      # For some reason `makeLibraryPath` on `cudatoolkit_11` maps to
-      # <cudatoolkit_11.lib>/lib which is different from <cudatoolkit_11>/lib.
-      patchelf --set-rpath "$rpath:${cudatoolkit}/lib:${lib.makeLibraryPath [ cudatoolkit.lib cudnn ]}" $file
+      patchelf --set-rpath "$rpath:${cudaLibPath}" $file
```
I believe there is now an `--add-rpath` available

samuela replied:
thanks for the pointer! done in https://github.com/NixOS/nixpkgs/pull/288829/commits/478d57de9fb84a69950fcb9876a63483108972af

GaetanLepage commented:
Result of `nixpkgs-review pr 288829` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

samuela commented:
I'll proceed to merge tomorrow unless there are any objections?

SomeoneSerge commented on pkgs/development/python-modules/jax/default.nix:
Oh. Is it absolutely necessary to move jax to the let-in and change the formatting?

samuela replied:
Iiuc this is necessary to use it in passthru. If there's another way I'm happy to change

SomeoneSerge replied:
Ah this. You could just ask for `jax` from `callPackage`. This has obvious implications for overrides (they have to be done in overlays if you care about passthru consistency) but this is a crime we commit through nixpkgs as far as I know

samuela replied:
Gotcha, ok done in https://github.com/NixOS/nixpkgs/pull/288829/commits/652924819ad676e9be337cafa4139f9b3b33a452!

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/bin.nix:
```diff
@@ -189,18 +189,12 @@ buildPythonPackage {
-      # For some reason `makeLibraryPath` on `cudatoolkit_11` maps to
-      # <cudatoolkit_11.lib>/lib which is different from <cudatoolkit_11>/lib.
-      patchelf --set-rpath "$rpath:${cudatoolkit}/lib:${lib.makeLibraryPath [ cudatoolkit.lib cudnn ]}" $file
+      patchelf --add-rpath "${cudaLibPath}" $file
```
`"$file"` (space splitting)

samuela replied:
done in https://github.com/NixOS/nixpkgs/pull/288829/commits/9e8cdafeb8387a8f2bc3179a12d93f184b2347cd

SomeoneSerge commented on pkgs/development/python-modules/jaxlib/bin.nix:
```diff
@@ -211,12 +205,14 @@ buildPythonPackage {
-    mkdir -p $out/bin
-    ln -s ${cudatoolkit}/bin/ptxas $out/bin/ptxas
+    mkdir -p $out/${python.sitePackages}/jaxlib/cuda/bin
+    ln -s ${lib.getBin cudaPackagesGoogle.cuda_nvcc}/bin/ptxas $out/${python.sitePackages}/jaxlib/cuda/bin/ptxas
```
I keep posting cosmetic suggestions and that feels wrong, but `lib.getExe' cudaPackagesGoogle.cuda_nvcc "ptxas"` 😆

samuela replied:
no, it's great! i'm learning new `lib` functions this way!

done in https://github.com/NixOS/nixpkgs/pull/288829/commits/9e8cdafeb8387a8f2bc3179a12d93f184b2347cd

samuela committed:
python3Packages.jaxlib-bin: cudatoolkit -> redist packages

samuela committed:
python3Packages.jax: add CUDA tests in passthru

Organize CUDA-enabled tests for jax and jaxlib into passthru scripts to facilitate testing.
